numpy/core/src/multiarray/item_selection.c

```diff
@@ -2128,6 +2129,82 @@ count_nonzero_bytes_384(const npy_uint64 * w)
     return r;
 }
 
+#if NPY_SIMD
+
+/* Count the zero bytes between `*d` and `end`, updating `*d` to point to where to keep counting from. */
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u8
+count_zero_bytes_u8(const npy_uint8 **d, const npy_uint8 *end, npy_uint8 max_count)
+{
+    const npyv_u8 vone = npyv_setall_u8(1);
+    const npyv_u8 vzero = npyv_zero_u8();
+
+    npy_intp lane_max = 0;
+    npyv_u8 vsum8 = npyv_zero_u8();
+    while (*d < end && lane_max <= max_count - 1) {
+        npyv_u8 vt = npyv_cvt_u8_b8(npyv_cmpeq_u8(npyv_load_u8(*d), vzero));
+        vt = npyv_and_u8(vt, vone);
+        vsum8 = npyv_add_u8(vsum8, vt);
+        *d += npyv_nlanes_u8;
+        lane_max += 1;
+    }
+    return vsum8;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u16
+count_zero_bytes_u16(const npy_uint8 **d, const npy_uint8 *end, npy_uint16 max_count)
+{
+    npyv_u16 vsum16 = npyv_zero_u16();
+    npy_intp lane_max = 0;
+    while (*d < end && lane_max <= max_count - 2*NPY_MAX_UINT8) {
+        npyv_u8 vsum8 = count_zero_bytes_u8(d, end, NPY_MAX_UINT8);
+        npyv_u16x2 part = npyv_expand_u16_u8(vsum8);
+        vsum16 = npyv_add_u16(vsum16, npyv_add_u16(part.val[0], part.val[1]));
+        lane_max += 2*NPY_MAX_UINT8;
+    }
+    return vsum16;
+}
+
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u32
+count_zero_bytes_u32(const npy_uint8 **d, const npy_uint8 *end, npy_uint32 max_count)
+{
+    npyv_u32 vsum32 = npyv_zero_u32();
+    npy_intp lane_max = 0;
+    while (*d < end && lane_max <= max_count - 2*NPY_MAX_UINT16) {
+        npyv_u16 vsum16 = count_zero_bytes_u16(d, end, NPY_MAX_UINT16);
+        npyv_u32x2 part = npyv_expand_u32_u16(vsum16);
+        vsum32 = npyv_add_u32(vsum32, npyv_add_u32(part.val[0], part.val[1]));
+        lane_max += 2*NPY_MAX_UINT16;
+    }
+    return vsum32;
+}
```

Comment:
```suggestion

```
I think there's no need for an extra block-level, two is enough plus the prev suggestion increased the iterations for u16 level